server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,378 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.histogram;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.util.CollectionUtil;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.lease.Releasables;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.DoubleArray;
+import org.elasticsearch.common.util.IntArray;
+import org.elasticsearch.index.fielddata.SortedNumericDoubleValues;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.BucketOrder;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.bucket.DeferableBucketAggregator;
+import org.elasticsearch.search.aggregations.bucket.DeferringBucketCollector;
+import org.elasticsearch.search.aggregations.bucket.MergingBucketsDeferringCollector;
+import org.elasticsearch.search.aggregations.bucket.nested.NestedAggregator;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+public class VariableWidthHistogramAggregator extends DeferableBucketAggregator {
+
+    private final ValuesSource.Numeric valuesSource;
+    private final DocValueFormat formatter;
+
+    private int numBuckets; // Final number of buckets (clusters) that should be produced by the coordinator node after reduction
+    private int shardSize; // Max number of buckets that this aggregator should return for reduction
+    private int cacheLimit; // Number of documents that are stored in memory before initial bucketing occurs
+    private int numCachedDocs;
+    private int avgBucketDistance;
+
+    final BigArrays bigArrays;
+    private DoubleArray cachedValues;
+    private DoubleArray clusterMaxes;
+    private DoubleArray clusterMins;
+    private DoubleArray clusterCentroids;
+    private IntArray clusterSizes; // Different than BucketsAggregator.docCounts: holds intermediate doc counts during clustering
```

Comment:
I wonder if it'd be cleaner to write it like:
```
private Work work;
```
and have two kinds of `Work`:
```
private class CacheValues extends Work {
  private DoubleArray cachedValues;
  private int numCachedDocs;
  @Override
  public void (LeafBucketCollector sub, int doc, double val) throws IOException {
    cachedValues = cachedValues.grow(bigArrays, numCachedDocs + 1);
    cachedValues.set(numCachedDocs, val);
    ....
```
and
```
private class MergeBuckets extends Work {
   private DoubleArray clusterMaxes;
   private DoubleArray clusterMins;
   private DoubleArray clusterCentroids;
  @Override
  public void (LeafBucketCollector sub, int doc, double val) throws IOException {
    int bucketOrd = getNearestBucket(val);
   ....
```

That way at a quick glance you could tell that the aggregator operates in two "modes". I haven't really thought it through, but I *think* something like this would be more readable to me.

Reply:
Of all of the comments I left that you just :+1:ed I didn't expect you to just :+1: this one. But, if you like the idea, great!

Reply:
It was a really good idea, I liked it! The code is much more readable now after I did this refactor.

I used 👍to indicate I implemented it btw. I didn't want to resolve your comments myself, in case you felt there was still work to do on them :)